src/lib/pubkey/frodokem/frodo_matrix.cpp

```diff
@@ -0,0 +1,513 @@
+/*
+ * FrodoKEM matrix logic
+ * Based on the MIT licensed reference implementation by the designers
+ * (https://github.com/microsoft/PQCrypto-LWEKE/tree/master/src)
+ *
+ * The Fellowship of the FrodoKEM:
+ * (C) 2023 Jack Lloyd
+ *     2023 René Meusel, Amos Treiber - Rohde & Schwarz Cybersecurity
+ *
+ * Botan is released under the Simplified BSD License (see license.txt)
+ */
+
+#include <botan/internal/frodo_matrix.h>
+
+#include <botan/assert.h>
+#include <botan/frodokem.h>
+#include <botan/hex.h>
+#include <botan/mem_ops.h>
+#if defined(BOTAN_HAS_AES)
+   #include <botan/internal/aes.h>
+#endif
```

Comment:
Larger question: is there really a reason

a) To include the AES support at all? It seems like everyone is moving to SHAKE for pqc

b) If AES support is required/useful (for standards conformance or contract reasons, etc): can we just make the AES support also mandatory? 99% of the time you want AES anyway for other reasons, and while you maybe don't need *Frodo-AES*, in the current approach there isn't any way to disable that anyway; if you enable plain old AES, plus Frodo, you get Frodo-AES. This contrasts with how Kyber was done where Kyber-90s is a distinct module that can be toggled off.

Reply:
> To include the AES support at all?

In contrast to the NIST drafts, currently the [ISO draft for (e)FrodoKEM](https://frodokem.org/files/FrodoKEM-standard_proposal-20230314.pdf) explicitly specifies the AES variant. Otherwise we would have left it out.

> This contrasts with how Kyber was done where Kyber-90s is a distinct module that can be toggled off.

Initially, we planned to build two sub-modules for FrodoKEM as well. However, in contrast to Kyber, AES in FrodoKEM is somewhat an "add-on", as both FrodoKEM-SHAKE and FrodoKEM-AES depend on SHAKE. 😞 So FrodoKEM without SHAKE doesn't make sense while FrodoKEM without AES still works partially.
You make a fair point, though, that we currently cannot switch off AES-support in FrodoKEM if both `frodokem` and `aes` are enabled in the build.

Let us try to extract the AES portions into a very thin module, see how much complexity that adds and then decide whether its worth it or not.

Reply:
> Let us try to extract the AES portions into a very thin module, see how much complexity that adds and then decide whether its worth it or not.

Done here: https://github.com/randombit/botan/pull/3679/commits/0694f1db15e0082450fe4b6727935e5f4ea86390
In my opinion the AES sub module doesn't bring much added complexity for giving users the flexibility to switch it off.

Reply:
For the sake of consistency, we could certainly do that. Note that you will need to explicitly enable `frodokem_shake` and `frodokem_aes`. There's (AFAIK) no concept of a "virtual" module feature that would enable both variants. Similarly, `kyber` enables the Keccak-based Kyber and `kyber_90s` enables the AES variant. There's an (internal) `kyber_common` with the base implementation that both flavors depend on.